File: src/data/http/retry.ts
/** Délai de la première nouvelle tentative. */
export const BASE_DELAY_MS = 500;

/**
 * Plafond. Sans lui, la huitième tentative attendrait déjà plus d'une minute et
 * la vingtième plus de cent heures : l'exponentielle n'est pas décorative.
 */
export const MAX_DELAY_MS = 30_000;

/**
 * Délai avant la tentative `attempt` (0-indexée), en millisecondes.
 *
 * La gigue est **injectée** plutôt que lue depuis `Math.random()` : c'est ce qui
 * rend la fonction testable, et une fonction de backoff non testée est une
 * fonction dont on ignore le comportement au moment où elle compte le plus.
 *
 * Pourquoi une gigue (C-15) : Hub'Eau n'annonce aucun quota chiffré, et
 * l'application n'a pas de proxy pour mutualiser la charge de sa base
 * installée. Sans étalement, tous les appareils réessaient à la même seconde
 * après une panne et achèvent un service public gratuit au moment précis où il
 * se relève.
 */
export function delayForAttempt(attempt: number, jitter: () => number = Math.random): number {
  const exponential = Math.min(BASE_DELAY_MS * 2 ** attempt, MAX_DELAY_MS);
  // La gigue s'ajoute au délai, puis on replafonne : sans ce second plafond
  // elle ferait dépasser MAX_DELAY_MS de 100 %.
  return Math.min(exponential + jitter() * exponential, MAX_DELAY_MS);
}

File: lib/data/http/retry.dart
// Recul exponentiel à gigue injectée (C-12) : aucune des APIs publiques du
// projet ne publie de quota chiffré, et aucun service intermédiaire ne
// mutualise la charge entre appareils. Sans étalement, tous les appareils
// réessaient à la même seconde après une panne partagée — la gigue casse
// cette synchronisation, et le tirage est injectable pour rester testable
// (le générateur de dart:math est autrement non déterministe).
import 'dart:math';

/// Base du recul avant gigue à la première tentative : 500 ms.
const Duration baseDelay = Duration(milliseconds: 500);

/// Plafond absolu du délai renvoyé, gigue comprise : 30 s.
const Duration maxDelay = Duration(seconds: 30);

/// Tentative au-delà de laquelle le décalage de bits déborderait avant même
/// d'avoir atteint le plafond. La base y vaut déjà directement le plafond
/// (`maxDelay / 2`) : la figer à ce cran ne change aucun résultat observable.
const int _maxShiftedAttempt = 30;

final Random _rng = Random();

/// Tirage uniforme dans `[0, 1)` via le générateur de `dart:math` — le seul
/// de tout le produit.
double _defaultJitter() => _rng.nextDouble();

/// Délai avant une nouvelle tentative HTTP, gigue comprise (C-12).
///
/// La base vaut `baseDelay × 2^attempt`, **plafonnée à `maxDelay / 2`**
/// (15 000 ms). Le délai renvoyé vaut `base + jitter() × base`, donc entre
/// la base seule (gigue nulle) et le double de la base (gigue maximale).
///
/// Le plafond porte sur la **base**, pas sur le résultat : plafonner le
/// résultat ferait valoir `min(30000 + gigue, 30000) = 30000` dès la
/// sixième tentative, et la gigue disparaîtrait précisément quand elle
/// compte le plus pour étaler la charge entre appareils. En plafonnant la
/// base, le résultat continue de varier entre `maxDelay / 2` et `maxDelay`
/// selon la gigue, même au plateau.
///
/// [attempt] négatif lève une [ArgumentError] — une tentative n'a pas de
/// rang négatif.
Duration delayForAttempt(
  int attempt, {
  double Function() jitter = _defaultJitter,
}) {
  if (attempt < 0) {
    throw ArgumentError.value(attempt, 'attempt', 'doit être positif ou nul');
  }

  final int shiftedAttempt = attempt > _maxShiftedAttempt
      ? _maxShiftedAttempt
      : attempt;
  final int rawBaseMs = baseDelay.inMilliseconds * (1 << shiftedAttempt);
  final int capMs = maxDelay.inMilliseconds ~/ 2;
  final int baseMs = rawBaseMs > capMs ? capMs : rawBaseMs;

  final double delayMs = baseMs + jitter() * baseMs;
  return Duration(milliseconds: delayMs.round());
}
